common/shared.c

```diff
@@ -234,6 +240,56 @@ uint64_t ft_info_to_mr_access(struct fi_info *info)
 	return mr_access;
 }
 
+uint64_t *ft_access_to_comb(uint64_t fixed_bits, uint64_t opt_bits)
+{
+	uint64_t flags[10];
+	int num_flags = 0;
+	uint64_t *combinations;
+	int buff_size;
+	int i;
+	int index;
+
+	for (i = 0; i < 64 && num_flags < 10; i++) {
+		if (opt_bits >> i & 1)
+			flags[num_flags++] = 1 << i;
+	}
+		
+	if (num_flags == 10)
+		return NULL;
+
+	buff_size = 1;
+	for (i = 0; i < num_flags; i++)
+		buff_size *= 2;
+
+	combinations = calloc(buff_size, sizeof(fixed_bits));
+
+	index = 0;
+	for (i = 1; i <= num_flags; ++i)
+		access_combinations(combinations, &index, flags, num_flags, i, 0, 0);
+
+	for (i = 0; i < buff_size - 1; i++)
+		combinations[i] |= fixed_bits;
+
+	return combinations;
+}
+
+void access_combinations(uint64_t *combinations, int *num_comb, uint64_t *flags,
+			 int num_flags, int choose, int curr, uint64_t building)
+{
+	if (choose + curr > num_flags)
+		return;
+
+	if (!choose) {
+		combinations[(*num_comb)++] = building;
+	} else {
+		for (; curr < num_flags; curr++) {
+			access_combinations(combinations, num_comb, flags, num_flags,
+					    choose - 1, curr + 1, building | flags[curr]);
+		}
+	}
+}
```

Comment:
I would look at converting the above two functions into more basic primitives.  Here's a rough example:

```c
#define for_each_bit(x, i)   for (i = 0; i < (8 << sizeof(x)); i++)
#define bit_isset(x, i)  (x >> i) & 1)

static inline int bit_set_count(u64 val)
{
    for_each_bit(val, i)
        if (bit_isset(val, i))
            cnt++;
    return cnt;
}

int ft_alloc_bit_combo(u64 fixed, u64 opt, u64 **combo, size_t *count)
{
    cnt = 1 + bit_set_count(opt);
    flags = calloc(cnt, sizeof *flags);

    index = 0;
    for_each_bit(opt, i)
        if (bit_isset(opt, i)
            flags[index++] = 1 << i;

    cnt = 1 << cnt;
    combo = calloc(cnt, sizeof u64);

    for (index = 0; index < cnt; index ++) {
        combo[index] = fixed;
        for_each_bit(index, i)
            combo[index] |= 1 << i;
    }
}

#define ft_free_bit_combo(combo)  free(combo)
```